### PR TITLE
cql3: SELECT CAST column names should match Cassandra's

### LINKED ISSUE #14508: SELECT CAST column names should match Cassandra's
When doing a `SELECT CAST(b AS int)`, Cassandra returns a column named `cast(b as int)`.
An application may depend on this specific column name to look at the results.
Currently, Scylla uses a different name - `system.castasint(b)`. For Cassandra compatibility, we should switch to the same name that Cassandra uses

@avikivity recently did the same for the `token` function to print exactly the same string as Cassandra prints for column names - see `operator<<` in  `cql3/expr/expression.cc`, the `is_token_function` (and we have a for_metadata case). 

Reproduced by a translated Cassandra unit test:
* `cassandra_tests/functions/cast_fcts_test.py`::`testNumericCastsInSelectionClause`.

I don't know if this name was a recent regression from the expressions rewrite, or we always used this different name. If we always used the different name, switching to Cassandra's may cause backward compatibility with our own old versions, so I don't know what to suggest (I think even in this case it's better to move towards Cassandra compatibility).

### PR DESCRIPTION
**Pull request by alexander-tur**
When doing a SELECT CAST(b AS int), Cassandra returns a column named cast(b as int).
An application may depend on this specific column name to look at the results.
Currently, Scylla uses a different name - system.castasint(b). For Cassandra compatibility, we should switch to the same name that Cassandra uses

fixes #14508

### CODE REVIEW

**Comment by denesb** [comment]
> @scylladb/scylla-maint please review and trigger CI for this pull request **CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14800

Kicked CI.

**Comment by avikivity** [comment]
Please use descriptive titles.

**Comment by nyh** [comment]
> Please use descriptive titles.

The title is now too long. It should be something that fits in one line (e.g., usually less than 80 characters) and then the rest of the commit log can be long.

**Comment by nyh** [comment]
> > Please use descriptive titles.
> 
> The title is now too long. It should be something that fits in one line (e.g., usually less than 80 characters) and then the rest of the commit log can be long.

Please also note that since this PR is a single commit, the title and the text of the pull request itself doesn't matter at all - only the commit itself matters, and its title, the full commit message, and the Fixes line should all be up-to-date inside the commit itself - not here in the PR.

**Comment by avikivity** [comment]
Please use `git log` to see how changelogs are typically written.

File: cql3/expr/expression.cc [hunk 1]
```diff
@@ -1136,8 +1136,15 @@ std::ostream& operator<<(std::ostream& os, const expression::printer& pr) {
                             } else if (!pr.for_metadata) {
                                 fmt::print(os, "{}({})", fn->name(), fmt::join(fc.args | transformed(to_printer), ", "));
                             } else {
-                                auto args = boost::copy_range<std::vector<sstring>>(fc.args | transformed(to_printer) | transformed(fmt::to_string<expression::printer>));
-                                fmt::print(os, "{}", fn->column_name(args));
+                                const sstring& fn_name = fn->name().name;
+                                if (!fn_name.find("castas")) {
```
[the remark this answers is not in the document]
Comment (nyh) [reply cql3/expr/expression.cc:1140]: If fn_name doesn't include the "system." (?) then maybe there's a different fn method to call to check that the keyspace name is "system".

File: test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py [hunk 2]
```diff
@@ -15,9 +15,9 @@ def testInvalidQueries(cql, test_keyspace):
     with create_table(cql, test_keyspace, "(a int primary key, b text, c double)") as table:
         assertInvalidMessage(cql, table, "cannot be cast to", "SELECT CAST(a AS boolean) FROM %s")
 
-# Marked cassandra_bug because of CASSANDRA-18647 (see comment below)
-@pytest.mark.xfail(reason="issue #14508")
-def testNumericCastsInSelectionClause(cql, test_keyspace, cassandra_bug):
+# cast to the same type should be optimised
+@pytest.mark.xfail(reason="a new issue should be opened")
+def testOptimiseNumericCast(cql, test_keyspace, cassandra_bug):
```
Comment (nyh) [inline test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py:20]: This test file was translated from Cassandra's unit tests - I don't want to rename the tests or start splitting them. This entire test was marked "xfail" because of issue 14508. If this was a mistake, and the test still fails after 14508 passes, where does it fail now? If there's a completely different bug, you should open a new issue - not write a comment "a new issue should be opened".

If you want to write a new short test just for 14508 and nothing else, please don't put it in this file (which should contain only tests translated from Cassandra) - rather put it in test/cql-pytest/test_cast_data.py. Thanks.

Comment (alexander-tur) [reply test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py:20]: @nyh  I guess Cassandra does some query optimizations and ignores cast to the same type. So CAST(int_col as int) is converted to int_col during the optimization and Cassandra returns just int_col as the column name.

Yes we can check the argument type in cast function and make a proper title here but I believe it's much better to create a new issue for query optimization.

btw could you please check https://github.com/scylladb/scylladb/issues/13250#issuecomment-1527104615

Comment (nyh) [reply test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py:20]: Yes, it's indeed possible to have a patch that solves 90% of a problem and create a new issue for the 10% left especially if looks not important and you think that solving the 100% in one time is too ugly or something. However, if you do that I'm still asking that you don't split or rename the tests in the `cassandra_tests` subdirectory, and instead create new tests specifically for the new issue, outside the `cassandra_tests` directory.

Comment (nyh) [reply test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py:20]: I can suggest two options:
1. You can fix the 10% remaining right now - if the cast is to the same type, drop it when printing the function name.
2. If you leave that last problem (cast to the same type's name), you can leave the test in cassandra_tests unmodified (it will sstill fail one the same issue...) and instead create two new tests in test/cql-pytest/ test_cast_data.py - one test will test the name of the cast to a different type and will pass, and the second test will test the name of the cast to the same time, and will xfail (and pass on Cassandra). If you do this, please note that the coding style in test_cast_data.py is different from the translated Cassandra tests. In particular, please try to use an existing table (we have "table1" there) or create a new shared table instead of creating a new table for each test.

Comment (alexander-tur) [reply test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py]: @nyh fixed

**Comment by denesb** [comment]
Kicked CI.

File: cql3/expr/expression.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexander-tur) [reply cql3/expr/expression.cc]: fixed

File: test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py (hunk 2 above)
Comment (alexander-tur) [reply test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py]: > I can suggest two options:
> 
> 1. You can fix the 10% remaining right now - if the cast is to the same type, drop it when printing the function name.

Hope I managed to optimize a query with cast to the same type. As result correct name is printed. Could you please check it?

**Comment by denesb** [comment]
@alexander-tur when updating the branch, please use the rebase option, not the merge option.

Re-kicked CI.

**Comment by denesb** [comment]
There are 3 failing dtests, I will look into fixing them.

**Comment by denesb** [comment]
Dtest fixes: https://github.com/scylladb/scylla-dtest/pull/3542

@alexander-tur while waiting for that to go in, please:
* drop the second commit
* change the title of the first commit to be shorter

@nyh please re-review.

**Comment by denesb** [comment]
Triggered CI. Dtest dependency was merged, so it should pass now (pending any flakyness).
@nyh please re-review.

**Comment by denesb** [comment]
Kicked CI. @nyh another ping.

**Comment by denesb** [comment]
@nyh ping, please check that your comments were addressed, or merge.

File: cql3/assignment_testable.hh [hunk 3]
```diff
@@ -32,7 +32,7 @@ public:
     }
 
     static bool is_exact_match(test_result tr) {
-        return tr != test_result::EXACT_MATCH;
+        return tr == test_result::EXACT_MATCH;
```
Comment (nyh) [inline cql3/assignment_testable.hh:35]: Wow, what a strange bug you fixed here :-)
Luckily, I checked and no existing code uses this function (that was written 9 years ago, commit 18d92f6cd64!) so it's not necessary to split this fix into a separate patch and open an issue for it, as we normally do.

File: test/cql-pytest/test_cast_data.py [hunk 4]
```diff
@@ -100,6 +100,21 @@ def test_cast_from_large_varint_to_varchar(cql, table1, cassandra_bug):
     cql.execute(f'INSERT INTO {table1} (p, cVarint) VALUES ({p}, {v})')
     assert [(str(v),)] == list(cql.execute(f"SELECT CAST(cVarint AS varchar) FROM {table1} WHERE p={p}"))
 
+# In test_cast_column_names we checked that columns are named in Cassandra style
```
Comment (nyh) [inline test/cql-pytest/test_cast_data.py:103]: It's confusing to mention a test's name and speak in past tense ("checked") about the test that **follows**. Note that other tests in this file we use this "In ... we checked" wording, but it refers to a **previous** test, and explains that this test is different in some way.
Also since this is a test that reproduces an issue, it's important to mention it (so if we ever see this test failing, we know which issue we forgot to backport, or regressed, etc.).
How about:
```python
# Test that when SELECTing a CAST, the column names returned have the expected
# format - "cast(x as int)", as Cassandra does. Reproduces issue #14508.
```

File: test/cql-pytest/test_cast_data.py [hunk 5]
```diff
@@ -100,6 +100,21 @@ def test_cast_from_large_varint_to_varchar(cql, table1, cassandra_bug):
...
+def test_cast_column_names(cql, table1, cassandra_bug):    
+    p = unique_key_int()
+    v = 32767456456456456456545678943512357658768763546575675
```
Comment (nyh) [inline test/cql-pytest/test_cast_data.py:106]: Nitpick: maybe use a small value (e.g., "7") to not mix this test with the issue of what happens when casting a huge varint to a small type? This matter has no relevance to this test.

File: test/cql-pytest/test_cast_data.py [hunk 6]
```diff
@@ -100,6 +100,21 @@ def test_cast_from_large_varint_to_varchar(cql, table1, cassandra_bug):
     cql.execute(f'INSERT INTO {table1} (p, cVarint) VALUES ({p}, {v})')
     assert [(str(v),)] == list(cql.execute(f"SELECT CAST(cVarint AS varchar) FROM {table1} WHERE p={p}"))
 
+# In test_cast_column_names we checked that columns are named in Cassandra style
+def test_cast_column_names(cql, table1, cassandra_bug):
```
Comment (nyh) [inline test/cql-pytest/test_cast_data.py:104]: Did you put here cassandra_bug by mistake?
cassandra_bug means that this test is known to fail on Cassandra. This shouldn't be the case here - the whole point of this issue was to make Scylla compatible with Cassandra, so the test for it should pass on Cassandra.
Please remove the "cassandra_bug" from this and the following test, and check that it works on Cassandra (`test/cql-pytest/run-cassandra test_cast_data.py::test_cast_column_names`).

File: test/cql-pytest/test_cast_data.py [hunk 7]
```diff
@@ -100,6 +100,21 @@ def test_cast_from_large_varint_to_varchar(cql, table1, cassandra_bug):
...
+    p = unique_key_int()
+    v = 32767456456456456456545678943512357658768763546575675
+    for t in ['int', 'double', 'text', 'tinyint', 'smallint']:
+        cql.execute(f'INSERT INTO {table1} (p, cVarint) VALUES ({p}, {v})')
+        assert cql.execute(f"SELECT CAST(cVarint as {t}) FROM {table1} WHERE p={p}").one()._fields[0] == f"cast_cvarint_as_{t}"
+
+#In test_ignore_cast_to_the_same_type we checked that query with casting of column to the same type is optimized
+def test_ignore_cast_to_the_same_type(cql, table1, cassandra_bug):    
+    p = unique_key_int()
+    v = 32767456456456456456545678943512357658768763546575675
```
Comment (nyh) [inline test/cql-pytest/test_cast_data.py:114]: Again, same comments as above regarding cassandra_bug and the unnecessarily-huge value for v.

File: test/cql-pytest/test_cast_data.py [hunk 8]
```diff
@@ -100,6 +100,21 @@ def test_cast_from_large_varint_to_varchar(cql, table1, cassandra_bug):
...
+    cql.execute(f'INSERT INTO {table1} (p, cVarint) VALUES ({p}, {v})')
+    assert cql.execute(f"SELECT CAST(p as int) FROM {table1} WHERE p={p}").one()._fields[0] == f"p"
```
Comment (nyh) [inline test/cql-pytest/test_cast_data.py:116]: nitpick: no need for the `f` in `f"p"`.

**Review by nyh — CHANGES_REQUESTED** [review]
Thanks, the code fix itself looks good, but I do have some requests regarding the test.

File: test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py [hunk 9]
```diff
@@ -15,8 +15,6 @@ def testInvalidQueries(cql, test_keyspace):
     with create_table(cql, test_keyspace, "(a int primary key, b text, c double)") as table:
         assertInvalidMessage(cql, table, "cannot be cast to", "SELECT CAST(a AS boolean) FROM %s")
 
-# Marked cassandra_bug because of CASSANDRA-18647 (see comment below)
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/cql-pytest/cassandra_tests/functions/cast_fcts_test.py:18]: I looked again, and you're right - this test will actually pass on recent versions Cassandra because this bug was fixed after I reported it. So you're indeed right to remove this line now. Thanks!

File: test/cql-pytest/test_cast_data.py [hunk 10]
```diff
@@ -100,6 +100,21 @@ def test_cast_from_large_varint_to_varchar(cql, table1, cassandra_bug):
     cql.execute(f'INSERT INTO {table1} (p, cVarint) VALUES ({p}, {v})')
     assert [(str(v),)] == list(cql.execute(f"SELECT CAST(cVarint AS varchar) FROM {table1} WHERE p={p}"))
 
+# In test_cast_column_names we checked that columns are named in Cassandra style
+def test_cast_column_names(cql, table1, cassandra_bug):    
+    p = unique_key_int()
+    v = 32767456456456456456545678943512357658768763546575675
+    for t in ['int', 'double', 'text', 'tinyint', 'smallint']:
+        cql.execute(f'INSERT INTO {table1} (p, cVarint) VALUES ({p}, {v})')
+        assert cql.execute(f"SELECT CAST(cVarint as {t}) FROM {table1} WHERE p={p}").one()._fields[0] == f"cast_cvarint_as_{t}"
```
Comment (nyh) [inline test/cql-pytest/test_cast_data.py:109]: It is sad that we need to check that version of the column name after it was botched ("cleaned") up by the Python driver (see _clean_column_names() in https://docs.datastax.com/en/drivers/python/2.7/_modules/cassandra/query.html), instead of the original name with spaces.
If you want to try something cool (and I don't insist, but please do try), you can consider telling the the driver to a different "factory", which doesn't "clean up" the column names. For example, something like (untested!):
```python
    saved_row_factory = cql.row_factory
    cql.row_factory = cassandra.query.dict_factory
    rows = cql.execute("SELECT CAST....")
    cql.row_factory = saved_row_factory
    assert rows[0].keys() == [f'cast(cvarint as {t}']
```

Comment (alexander-tur) [reply test/cql-pytest/test_cast_data.py:109]: Hello @nyh, I've tried this and got "ValueError: Cannot set Session.row_factory while using Configuration Profiles. Set this in a profile instead". I'll try to find a solution but not sure about the time. Actually as far as cassandra test covers this case we don't need this test at all.

Comment (nyh) [reply test/cql-pytest/test_cast_data.py:109]: You can feel free to ignore this idea - I'll do it later (just because I think it's cool). The test we already have, with those silly underscores that the Python driver adds, is good enough already. Just please look at my other comments, and I'll be happy to merge the next version.

Comment (alexander-tur) [reply test/cql-pytest/test_cast_data.py:109]: @nyh , Done

**Comment by denesb** [comment]
> @scylladb/scylla-maint please review and trigger CI for this pull request **CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14800

CI kicked.

**Comment by avikivity** [comment]
Please change the title of the pull request so it is concise. Not the title becomes part of the merge commit, see `git log` for the style we use.
